llm/llm.go

```diff
@@ -50,92 +50,88 @@ func New(model string, adapters, projectors []string, opts api.Options) (LLM, er
-	kv := 2 * 2 * int64(opts.NumCtx) * int64(ggml.KV().BlockCount()) * int64(ggml.KV().EmbeddingLength()) * int64(ggml.KV().HeadCountKV()) / int64(ggml.KV().HeadCount())
+	// fp16 k,v = (1 (k) + 1 (v)) * sizeof(float16) * n_ctx * n_layer * n_embd / n_head * n_head_kv
+	kv := 2 * 2 * int64(opts.NumCtx) * int64(ggml.KV().BlockCount()) * int64(ggml.KV().EmbeddingLength()) / int64(ggml.KV().HeadCount()) * int64(ggml.KV().HeadCountKV())
 
 	// this amount is the overhead + tensors in memory
 	// TODO: get this from the llama.cpp's graph calculations instead of
 	// estimating it's 1/6 * kv_cache_size * num_gqa
 	graph := int64(ggml.KV().GQA()) * kv / 6
+	mem += graph
 
-	if slices.Contains(cpuOnlyFamilies, ggml.KV().Architecture()) {
+	if mem > vram {
 		opts.NumGPU = 0
 	}
 
-	info := gpu.GetGPUInfo()
-	switch runtime.GOOS {
-	case "darwin":
-		if opts.NumGPU == 0 {
-			break
-		}
+	if opts.NumGPU < 0 && info.Library != "cpu" {
+		for opts.NumGPU = 0; opts.NumGPU < int(ggml.KV().BlockCount()) && vram > mem; opts.NumGPU++ {
+			layer := ggml.LayerSize(fmt.Sprintf("blk.%d.", opts.NumGPU))
+			layerKV := kv / int64(ggml.KV().BlockCount())
+			if vram <= mem+layer+layerKV {
+				break
+			}
 
-		if size+kv+graph > vram {
-			slog.Info("not enough vram available, setting num_gpu=0")
-			opts.NumGPU = 0
-			break
+			mem += layer + layerKV
 		}
 
-		// TODO: implement layer splitting on macOS
-		opts.NumGPU = 999
-	default:
-		if info.Library == "cpu" {
-			slog.Info("GPU not available, falling back to CPU")
-			opts.NumGPU = 0
-			break
+		// only offload output tensors if the repeating tensors have been offloaded
+		if layer := ggml.LayerSize("output."); opts.NumGPU >= int(ggml.KV().BlockCount()) && vram > mem+layer {
+			opts.NumGPU++
+			mem += layer
 		}
 
-		// don't use GPU at all if no layers are loaded
-		if opts.NumGPU == 0 {
-			info.Library = "cpu"
-			info.Variant = gpu.GetCPUVariant()
-			break
-		}
+		opts.VRAMUsed = mem
+		opts.VRAMTotal = vram
+	}
 
-		// user-defined GPU count
-		if opts.NumGPU != -1 {
-			break
-		}
+	if opts.NumGPU == 0 || slices.Contains(cpuOnlyFamilies, ggml.KV().Architecture()) {
+		info.Library = "cpu"
+	}
 
-		// the "main" GPU needs the most memory and determines the limit
-		// of how many layers can be loaded. It needs to fit:
-		// 1. the full compute graph allocation for all devices (graph)
-		// 2. the proportional kv cache for all devices (kv * % layers)
-		// 3. the proportional model (size * % layers / # devices)
-		// This estimates the number of layers
-		maxlayers := int64(ggml.KV().BlockCount()) + 1
-		devices := int64(info.DeviceCount)
-		avg := vram / devices
-		layers := maxlayers * (avg - graph) / (kv + size/devices)
-		if layers > maxlayers {
-			layers = maxlayers
-		}
+	slog.Info("llm", "model", model, "adapters", adapters, "projectors", projectors, "opts", opts, "info", info)
+	slog.Info("mem", "mem", mem, "vram", vram, "kv", kv, "graph", graph, "layers", opts.NumGPU)
```

Comment:
These sort of feel more like debug statements, but I totally see why setting these at info in the short-term makes sense in case we have any problems, so we can quickly spot where things went sideways on the calculations.  If that was your intent, maybe a comment breadcrumb to remind us to quiet this down to debug level after a few releases?  If your intent was to keep them around longer, maybe expand the message so it's more actionable/understandable for users.